### PR TITLE
Improve stdin parsing

### PR DESCRIPTION
**Pull request by GregorCH**

### CODE REVIEW

File: ipet/parsing/ReaderManager.py [hunk 1]
```diff
@@ -363,13 +393,11 @@ def fromXMLFile(xmlfilename):
 
     @staticmethod
     def processXMLElem(elem):
+        # FARI1 what if not? rm undefined?
```
Comment (GregorCH) [inline ipet/parsing/ReaderManager.py]: If the element is not a reader manager, this method throws a "ReferencedBeforeAssignment" error at the return.

File: ipet/parsing/ReaderManager.py [hunk 2]
```diff
@@ -296,46 +302,71 @@ def collectData(self):
         """
         assert(self.testrun != None)
 
-        # sort the files by context
-        self.filestrings = sorted(self.filestrings, key = lambda x:self.sortingKeyContext(self.filenameGetContext(x)))
+        # sort the files by context, for example: outfiles should be read before solufiles
+        if not self.testrun.inputfromstdin:
```
Comment (GregorCH) [inline ipet/parsing/ReaderManager.py]: By the way, self.testrun.inputfromstdin references a member of a different class --> We would need a method here.

File: ipet/parsing/Solver.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+                    self.addData(Key.PrimalBound, pb)
+            
+    def addData(self, key, data):
+        """
+        adds data to local data-dictionary
+        """
+        # FARIDO is it okay to overwrite data?
+        self.data[key] = data
+    
+    def readline(self, line):
```
Comment (GregorCH) [inline ipet/parsing/Solver.py]: camel case the "l"

File: ipet/parsing/Solver.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+from ipet.misc import misc
+from ipet.parsing.Key import Key
+from .StatisticReader import StatisticReader
+
+class Solver(StatisticReader):
+    
+    DEFAULT=None
+    DEFAULT_SOLVERID="defaultSolver"
+    
+    def __init__(self,
```
Comment (GregorCH) [inline ipet/parsing/Solver.py]: The constructor could be simplified by using groups and regular expressions as follows, I exemplify this for the primal bound. Currently, a pattern, and an index are used to distinguish the primal bound. It would be much simpler to use the following, combined expression:

```
inputstr = "Primal Bound       : +3.02000000000000e+02 (1 solutions)"
import re
primalboundexpression = re.compile("Primal Bound       : (\S+)")
print primalboundexpression.match(inputstr).groups()

>>> ('+3.02000000000000e+02',)

```

The design would be the following: The constructor receives patterns (strings) or None, and compiles expressions such as the above.

The method extractPrimalBound() would be changed to
```
try:
     pb = float(self.primalboundexpression.groups()[0])
except:
     pb = None
```
In a nutshell:

- Let's base parsing on simpler patterns such as the above to simplify the constructor, otherwise we drown in argument checking hell.

- Subclasses should also directly specify the expressions as static class member like so:

```
class SCIPSolver(Solver):

     primalbound_expression = re.compile("Primal Bound       : (\S+)")
     dualbound_expression = re.compile(...)
     ... remaining expressions...

    def __init__(self, primalbound_pattern=None):
         if primalbound_pattern is not None:
              self.primalbound_expression = re.compile(primalbound_pattern)
         ...
```

File: ipet/parsing/ReaderManager.py (hunk 2 above)
Comment (fschloesser) [inline ipet/parsing/ReaderManager.py]: What about all the magic with reading settings and parameters from filenames then? 
Is that a problem if we have a mix of .out- and .solu-files?

Comment (fschloesser) [inline ipet/parsing/ReaderManager.py]: yes!

File: ipet/parsing/ReaderManager.py [hunk 5]
```diff
@@ -177,108 +184,107 @@ def registerListOfReaders(self, readers):
...
 
-    def endOfInstanceReached(self, line):
+    def endOfProblemReached(self, line):
         if line.startswith(self.problemendexpression):
             return True
         else:
             return False
 
+    def startOfProblemReached(self, line):
+        if line.startswith(self.problemexpression):
```
Comment (fschloesser) [inline ipet/parsing/ReaderManager.py]: yes!

File: ipet/parsing/ReaderManager.py (hunk 1 above)
Comment (fschloesser) [inline ipet/parsing/ReaderManager.py]: okay

File: ipet/parsing/ReaderManager.py (hunk 2 above)
Comment (GregorCH) [inline ipet/parsing/ReaderManager.py]: Reading something from a filename should soon be deprecated IMHO. All of this, especially the settings, are what I consider Metainformation. Meta information is not relevant in itself, but for combining runs in an experiment, it might be relevant. My idea with Meta-Information is the following:

It can either be specified in an outfile directly as follows:

\@Settings default
\@Seed 5
\@Machine opt55


Or in a file with the same basename and the .meta extensions as follows, e.g., check.[blablabla].meta:
"""
Settings default
Seed 5
Machine opt55
"""
This, however, is not part of this review.
or, alternatively, in the outfile directly, as follows:

File: ipet/parsing/Solver.py (hunk 4 above)
Comment (fschloesser) [inline ipet/parsing/Solver.py]: good!

File: ipet/parsing/ReaderManager.py (hunk 2 above)
Comment (fschloesser) [inline ipet/parsing/ReaderManager.py]: that'll not make it easier finding out the solver, or do you have a good idea about that?
